ares/n64/cpu/recompiler.cpp

```diff
@@ -9,16 +10,68 @@ auto CPU::Recompiler::pool(u32 address) -> Pool* {
   return pool;
 }
 
-auto CPU::Recompiler::block(u64 vaddr, u32 address, bool singleInstruction) -> Block* {
-  if(auto block = pool(address)->blocks[address >> 2 & 0x3f]) return block;
-  auto block = emit(vaddr, address, singleInstruction);
-  if(block) {
-    pool(address)->blocks[address >> 2 & 0x3f] = block;
+auto CPU::Recompiler::blockCacheKey(u32 address, u32 state) -> u64 {
+  return u64(address) | (u64(state) << 32);
+}
+
+auto CPU::Recompiler::block(u64 vaddr, u32 address) -> Block* {
+  u32 state = jitContext.stateBits;
+  u32 idx = (address >> 2) & 0x3f;
+  assert(idx < sizeof(Pool::rows)/sizeof(Pool::rows[0]));
+
+  {
+    Pool* p = pool(address);
+    if (auto row = p->rows[idx]) {
+      if (row->stateBits == state) {
+        assert(row->block);
+        return row->block;
+      } else {
+        auto blockIt = blockCache.find(blockCacheKey(address, state));
```

Comment:
This is reusing any previously compiled block with the same start address and the same CPU context. However, the instructions could be entirely different! This is why the RSP recompiler incorporates a hash of the instructions in the lookup.

Reply:
Yikes. Yeah I forgot to clean up the block cache in `invalidatePool(address)`. I added an inverse mapping from a pool index to a list of cache keys but it's kind of cheesy. I'm not really happy with it. I wonder if there are other possible solutions other than a block content hash and this.

Reply:
Yikes. Yeah I forgot to clean up the block cache in `invalidatePool(address)`. I added an inverse mapping from a pool index to a list of cache keys but it's kind of cheesy. I'm not really happy with it. I wonder if there are other possible solutions other than a block content hash and this.